src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonDerivedTypeAttribute.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// When placed on a type declaration, indicates that the specified subtype should be opted into polymorphic serialization.
+    /// </summary>
+    [System.Runtime.Versioning.RequiresPreviewFeatures("Polymorphic serialization is currently in preview.")]
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Interface, AllowMultiple = true, Inherited = false)]
+    public class JsonDerivedTypeAttribute : JsonAttribute
+    {
+        /// <summary>
+        /// Initializes a new attribute with specified parameters.
+        /// </summary>
+        /// <param name="derivedType">A derived type that should be supported in polymorphic serialization of the declared based type.</param>
+        public JsonDerivedTypeAttribute(Type derivedType)
```

Comment:
That scenario is very much something we need to support. To answer your question, it is meant to be handled using the metadata model. Hadn't thought about specifying the derived type by string, although I would imagine it must be awkward to do (should require either full name or assembly qualified name and would break when refactoring happens). Do you think it might be helpful to users?

Reply:
I can see scenario where users might want to extend existing set of polimorphic types. IMO, something like `[JsonAllowDerivingFrom(typeof(BaseClass), "myId")]` on the new class. This will be possible through contract resolver though so we might want to wait with adding such attribute